asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go

```diff
@@ -139,107 +142,86 @@ func (t *SimpleChaincode) CreateAsset(ctx contractapi.TransactionContextInterfac
 	if err != nil {
 		return err
 	}
-	//  Save index entry to state. Only the key name is needed, no need to store a duplicate copy of the asset.
+	//  Save index entry to world state. Only the key name is needed, no need to store a duplicate copy of the asset.
 	//  Note - passing a 'nil' value will effectively delete the key from state, therefore we pass null character as value
 	value := []byte{0x00}
 	return ctx.GetStub().PutState(colorNameIndexKey, value)
 }
 
-// ReadAsset - read a asset from chaincode state
-func (t *SimpleChaincode) ReadAsset(ctx contractapi.TransactionContextInterface, ID string) (*asset, error) {
-	assetJSON, err := ctx.GetStub().GetState(ID)
+// ReadAsset retrieves an asset from the ledger
+func (t *SimpleChaincode) ReadAsset(ctx contractapi.TransactionContextInterface, assetID string) (*asset, error) {
+	assetBytes, err := ctx.GetStub().GetState(assetID)
 	if err != nil {
-		return nil, err
-	} else if assetJSON == nil {
-		return nil, fmt.Errorf("%s does not exist", ID)
+		return nil, fmt.Errorf("failed to get asset %s: %v", assetID, err)
+	}
+	if assetBytes == nil {
+		return nil, fmt.Errorf("asset %s does not exist", assetID)
 	}
 
-	asset := new(asset)
-	err = json.Unmarshal(assetJSON, asset)
+	var asset *asset
+	err = json.Unmarshal(assetBytes, &asset)
 	if err != nil {
 		return nil, err
 	}
 
 	return asset, nil
 }
 
-// DeleteAsset - remove a asset key/value pair from state
+// DeleteAsset removes an asset key-value pair from the ledger
 func (t *SimpleChaincode) DeleteAsset(ctx contractapi.TransactionContextInterface, assetID string) error {
-
-	// to maintain the color~name index, we need to read the asset first and get its color
-	assetJSON, err := ctx.GetStub().GetState(assetID) //get the asset from chaincode state
+	asset, err := t.ReadAsset(ctx, assetID)
 	if err != nil {
-		return fmt.Errorf("Failed to get state for %s", assetID)
-	} else if assetJSON == nil {
-		return fmt.Errorf("Asset does not exist %s", assetID)
-	}
-
-	var asset asset
-	err = json.Unmarshal([]byte(assetJSON), &asset)
-	if err != nil {
-		return fmt.Errorf("Failed to decode JSON of %s", assetID)
+		return err
 	}
 
-	err = ctx.GetStub().DelState(assetID) //remove the asset from chaincode state
+	err = ctx.GetStub().DelState(assetID)
 	if err != nil {
-		return fmt.Errorf("Failed to delete state:" + err.Error())
+		return fmt.Errorf("failed to delete asset %s: %v", assetID, err)
 	}
 
-	// maintain the index
 	indexName := "color~name"
 	colorNameIndexKey, err := ctx.GetStub().CreateCompositeKey(indexName, []string{asset.Color, asset.ID})
 	if err != nil {
-		return fmt.Errorf(err.Error())
+		return err
 	}
 
-	//  Delete index entry to state.
```

Comment:
keeping a comment will help a reader

`// Delete the colorName Index Key for this Asset`

Reply:
Agreed, I realize now it was an index deletion, not an asset deletion. In general code that is obvious should not be commented, but the fact this is an index deletion, the comment should remain to emphasize this wasn't a simple asset deletion.

Reply:
done